### ISSUE TITLE
All WhateverCode's are `eqv` to each other

### ISSUE THREAD

**Issue by Leont** [rakudo/rakudo#5116]
## The Problem
```
$ raku -e 'my $a = *.Int; my $b = *.Str; say $a eqv $b'
True
```

## Expected Behavior
`$a` and `$b` are not equivalent because they don't do the same thing

## Actual Behavior
They are considered equivalent

## Environment
```
Welcome to Rakudo™ v2022.07.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2022.07.
```

**Comment by 2colours** [comment]
The source of the problems: https://github.com/rakudo/rakudo/blob/2022.07/src/core.c/Mu.pm6#L1321

The `.raku` string representations are being compared. For `WhateverCode`, it's always just `WhateverCode.new`, hence the unwanted result.

**Comment by lizmat** [comment]
FWIW, this problem (if it is a problem) is more general for any class that doesn't provide a specific `.raku` method:
```raku
class A { }
dd A.new eqv A.new;  # True
```

**Comment by lizmat** [comment]
I think therefore this should be fixed by adding a specific `WhateverCode.raku` method, that will be different for different `WhateverCode`s.

**Comment by Leont** [comment]
> FWIW, this problem (if it is a problem) is more general for any class that doesn't provide a specific .raku method:

Any class that doesn't that doesn't (only?) have public attributes.

**Comment by lizmat** [comment]
Well, any class with public attributes and given the same values, would produce the same .raku, but that still could mean they're not equivalent.

**Comment by 2colours** [comment]
> Well, any class with public attributes and given the same values, would produce the same .raku, but that still could mean they're not equivalent.

I would say that isn't necessarily a bad thing, although perhaps it would be nicer to have the `===` semantics as the default and a provided `raku` overload the specific case.

**Comment by Leont** [comment]
> I would say that isn't necessarily a bad thing, although perhaps it would be nicer to have the === semantics as the default and a provided raku overload the specific case.

I think it's very useful behavior for classes where all members are public (e.g. one's stereotypical `class Point`), and breaking that expectation would break a bunch of code. That said for classes that are partly or entirely private I can see your argument.

**Comment by 2colours** [comment]
Apparently this fix had to be reverted for the upcoming Rakudo release.

**Comment by jubilatious1** [comment]
By @Leont 's logic, all below should be True:
```
~ % raku -e 'my $a = *.IntStr; my $b = *.IntStr; say $a eqv $b'
True
~ % raku -e 'my $a = *.IntStr; my $b = *.IntStr; say $a eq $b'
WhateverCode object coerced to string (please use .gist or .raku to do that)
  in block <unit> at -e line 1
WhateverCode object coerced to string (please use .gist or .raku to do that)
  in block <unit> at -e line 1
True
~ % raku -e 'my $a = *.IntStr; my $b = *.IntStr; say $a == $b'
Cannot resolve caller Numeric(WhateverCode:D: ); none of these signatures matches:
    (Mu:U \v: *%_)
  in block <unit> at -e line 1
```

**Comment by jubilatious1** [comment]
@2colours But it's called `.IntStr` coercion... .

**Comment by 2colours** [comment]
@jubilatious1 the return value, yes. But you compared the functions themselves. The functions are `WhateverCode` and they could indeed contain "whatever code" so trying to treat them as a number or string is not really a good idea.

**Comment by jubilatious1** [comment]
I compared the ___functions___?  How is coercion to a Class/Type different from declaring that Class/Type from the get-go?  Is above not the equivalent to:

```raku
~ % raku -e 'my IntStr $a; my IntStr $b; say $a eqv $b'
True
~ % raku -e 'my IntStr $a; my IntStr $b; say $a eq $b'
Invocant of method 'Str' must be an object instance of type
'Allomorph', not a type object of type 'IntStr'.  Did you forget a
'.new'?
  in block <unit> at -e line 1

~ % raku -e 'my IntStr $a; my IntStr $b; say $a == $b'
Use of uninitialized value of type IntStr in numeric context
  in block <unit> at -e line 1

~ % raku -e 'my IntStr $a; my IntStr $b; say $a ~~ $b'
True
~ % raku -e 'my IntStr $a; my IntStr $b; say $a === $b'
True
```

@Leont

**Comment by 2colours** [comment]
No, it's not even remotely similar to that code. `*.IntStr` is more or less the same as `{ .IntStr }` as a block that uses `$_` as its parameter, or `-> $x { $x.IntStr }` as a pointy block. It takes an argument and if you call it with an argument, it will call the IntStr method on it, which is at this point just an arbitrary method. Just look at [this type chart](https://docs.raku.org/type/WhateverCode#typegraphrelations).

**Comment by jubilatious1** [comment]
From the Docs:

>...subexpressions may impose their own Whatever star rules:
```raku
my @a = (0, 1, 2);
say get-val($c, @a[*-1]) # 2, because the star belongs to the Array class 
```
>

When `$_` is unset you're trying to say that `*` Whatever-star is it's own function. But according to the doc _"subexpressions may impose their own Whatever star rules"_. When a `*` Whatever-star is placed in a subexpression that is clearly a comparison, why not use in as an anonymous coercer? The user appears to be coercing to a class to do a comparison anyway, then chooses an operator that's appropriate.

**Comment by vrurg** [comment]
@jubilatious1 You're mixing up two different terms.

A `Whatever` or "whatever star" is a plain `*` which, roughly, indicated "match anything". Not always, but that depends on how code interprets it. I.e. code can treat it as infinity as in ranges.

Contrary to that, any _expression_ which is using an asterisk is a `WhateverCode`. Try that:

```
sub foo(\expr) { say expr.WHICH }
foo(*);
foo(*-1);
```

Basically, in your example with array `*-1` is an expression, thus it's a `WhateverCode` and the entire thing unwinds into something like `@a[(*-1).(@a.elems)]`.

**Comment by 2colours** [comment]
> ...subexpressions may impose their own Whatever star rules:

I don't know what this means.

> ```raku
> my @a = (0, 1, 2);
> say get-val($c, @a[*-1]) # 2, because the star belongs to the Array class 
> ```

We have talked about a lot of Raku on multiple platforms, regarding various matters. An essential thing that I'm trying to teach you as a rule of thumb: most clever Raku stuff is _not_ syntax but special handling of certain first-class types. How does it apply here? `WhateverCode` is essentially like a function, and `* - 1` is essentially a function that subtracts one from its argument. You should suspect that you can pass in _any subtractor function_ and have the same "counting backwards from the array length" behavior.

```raku
sub minus-one($x) { $x - 1 }
my @a = (0, 1, 2);
say @a[&minus-one]; # 2, because when you index an array with a single-parameter function, it will be called with the length of the array as the argument, and then the array will be indexed with the result
```

> When `$_` is unset you're trying to say that `*` Whatever-star is it's own function.

Perhaps I didn't express myself clear enough but that's definitely not what I was trying to say. When the Whatever-star appears in an expression, two things can happen: either the will be passed along like normal values are (this happens when you write `@a[*]` or `*.WHAT`), or a code object - basically a function - will be constructed from the whole expression where WhateverStars are replaced by positional parameters (this happens with `* - 1` inside `@a[* - 1]` or `*.IntStr`).

> When a `*` Whatever-star is placed in a subexpression that is clearly a comparison

This is the last part that I still understood, and it's not correct. The moment you started typing `.IntStr` after the asterisk, there was no Whatever-star to be talked about. There was only a `WhateverCode` object, which is, again, kind of a function. The comparison happens between these stored function instance.

**Comment by ab5tract** [comment]
This is fixed in RakuAST:

```
> set-rakuast; r 'my &a = *.uc; my &b = *.lc; dd &a eqv &b; dd &a eqv my &c = *.uc'
Bool::False
Bool::True
```

**Comment by Leont** [comment]
> This is fixed in RakuAST

Awesome!
